### ISSUE TITLE
Regexp unable to distinguish upper and lower case ASCII characters

### ISSUE THREAD

**Issue by Kratacoa** [newsboat/newsboat#1197]
As far as I know, this is by design, but I don't understand the reason behind it.
One use case that would be useful is using only uppercase characters in empty links used as delimiters for different feeds in the `urls` file.

**Comment by Minoru** [comment]
I wasn't around back than, but apparently it was so from day 1:

* `=~` and `!~` in filters, e.g. `ignore-article`: https://github.com/newsboat/newsboat/commit/79adeedd44277c41793ca0e31ede89c1778b818a
* `highlight` and `highlight-article`: https://github.com/newsboat/newsboat/commit/f647d92a84c0cdc00b660a32b70d0528a49bb3f0

No idea why, though.

> One use case that would be useful is using only uppercase characters in empty links used as delimiters for different feeds in the urls file.

I understand what delimiters you're talking about (#412), but I don't get why case would play any role here. Can you give me an example? I guess I need your `highlight` rules and a part of your urls file.

As for the underlying issue, a short-term fix might be a new setting that instructs `Matcher` and `RegexManager` to add or omit `REG_ICASE` flag when compiling the regex (with a caveat that changing a setting during runtime won't affect already-compiled regexes). This is blocked by #631 and by `ConfigContainer` being unavailable from Rust (I'll file an issue about that).

**Comment by Minoru** [comment]
Oh, and a long-term solution would be to add flags to regexes, like `/regex/i` in Perl. This will also make the previous caveat obvious: if you didn't specify a flag, or specified a different one, and now you can't change the regex (because e.g. you can't change `highlight` rules in runtime), the decision is fixed and you can't affect it.

**Comment by Kratacoa** [comment]
> I understand what delimiters you're talking about (#412), but I don't get why case would play any role here. Can you give me an example? I guess I need your highlight rules and a part of your urls file.  

`highlight all "^[A-Z]+$" red` for highlighting `MATHEMATICS`.
Currenly in my [urls file](https://github.com/Kratacoa/dotfiles/blob/master/config/newsboat/.config/newsboat/urls) I use a fancy thick dash to highlight such delimiters, but I would prefer to have a more noticeable alternative.

The issue you linked seems to be a more appropriate place to discuss my use case, but maybe a more "normal" regexp implementation will cover some other situations I can't envision.

**Comment by Minoru** [comment]
Ah, got it. I'm afraid that your approach will still have false positives even when Newsboat starts taking case into account; for example, your regex would highlight "RSS" and "USA", which aren't your labels.

The only real solution for that is #412 (so Newsboat "knows" which lines are delimiters and which ones are text, and can highlight them differently), but in the meantime, you can leverage your thick dash to disambiguate between them: `highlight all "^[A-Z]+ ─+$" red`.

Радий зустріти співвітчизника ^_^ Якщо користуєшся українською локалізацією і побачиш помилку — пиши, а краще надсилай патч (усе в файлі po/uk.po). Я, чесно кажучи, мало спілкуюсь українською, а в Newsboat взагалі використовую англійську, тому локалізація може бути не дуже вдалою.

**Comment by Kratacoa** [comment]
> Ah, got it. I'm afraid that your approach will still have false positives even when Newsboat starts taking case into account; for example, your regex would highlight "RSS" and "USA", which aren't your labels.

That is true, but can be corrected through ad hoc names for feeds; still, a proper delimiter would be a better solution.

> The only real solution for that is #412 (so Newsboat "knows" which lines are delimiters and which ones are text, and can highlight them differently), but in the meantime, you can leverage your thick dash to disambiguate between them: `highlight all "^[A-Z]+ ─+$" red`.

I've tried but apparently it doesn't support that character at all :/
How rich is the set of characters supported by the regex interpeter?

> Радий зустріти співвітчизника ^_^ Якщо користуєшся українською локалізацією і побачиш помилку — пиши, а краще надсилай патч (усе в файлі po/uk.po). Я, чесно кажучи, мало спілкуюсь українською, а в Newsboat взагалі використовую англійську, тому локалізація може бути не дуже вдалою.

Я так само користуюсь англійською. Постараюсь на днях попробувати і українську, якщо замічу недостатки однозначно поправлю!

**Comment by Minoru** [comment]
Ah, my bad: it should be `highlight all "[A-Z]+ ─+$" red`, without `^` at the beginning of the regex. The regex matches the whole line, and since the line in a feedlist starts with spaces and then some digits `^[A-Z]+` never matches. I verified that this updated line works with the urls file you linked earlier.

The regex is processed by `regcomp()` and `regexec()`, so I'd expect them to support everything provided that:
1. your locale matches the encoding of your urls file; and
2. the characters you use do not contain zero byte in your locale's encoding.

**Comment by Kratacoa** [comment]
> Ah, my bad: it should be highlight all "[A-Z]+ ─+$" red, without ^ at the beginning of the regex. The regex matches the whole line, and since the line in a feedlist starts with spaces and then some digits ^[A-Z]+ never matches. I verified that this updated line works with the urls file you linked earlier.

It does indeed work, thanks! Lovely :)
